File: CS.NET/PdfViewerCSharpAPI_R2/DocumentManagement/AlwaysRememberCache.cs
// -----------------------------------------------------------------------
// <copyright file="DummyCache.cs" company="">
// TODO: Update copyright text.
// </copyright>
// -----------------------------------------------------------------------

namespace PdfTools.PdfViewerCSharpAPI.DocumentManagement
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;
    using Requests;
    using PdfTools.PdfViewerCSharpAPI.Utilities;



    /// <summary>
    /// TODO: Update summary.
    /// </summary>
    public class AlwaysRememberCache<Arguments, ObjectToCache, Parametrization> : IGenericCache<Arguments, ObjectToCache, Parametrization>
    {

        private IDictionary<Arguments, ObjectToCache> dict;
        private ISet<Arguments> guessSet;
        private ISet<Arguments> exactlyLoadedSet;
        private Dictionary<Arguments, APdfRequest<Arguments, ObjectToCache>> pendingRequests; //arguments that are loading asynchronously and have not returned yet
        private Parametrization param;
        private int numPagesLoadedOnLastGuessGeneration = 0;
        private Object dictLock = new Object();

        public AlwaysRememberCache()
        {
            dict = new Dictionary<Arguments, ObjectToCache>();
            guessSet = new HashSet<Arguments>();
            exactlyLoadedSet = new HashSet<Arguments>();
            pendingRequests = new Dictionary<Arguments, APdfRequest<Arguments, ObjectToCache>>();
        }
        public AlwaysRememberCache(LoadObjectAsync<Arguments, ObjectToCache> loadObjectDelegate)
            : this()
        {
            this.loadObjectAsyncDelegate = loadObjectDelegate;
        }

        public ObjectToCache GetGuess(Arguments arguments)
        {
            lock (dictLock)
            {
                if (dict.Count <= 0)
                    return Get(arguments);//we cant guess anything when the dict is empty

                if (!dict.ContainsKey(arguments))
                {
                    //if we have loaded a lot more pages since we last generated a new guess, we should invalidate that old guess and generate a new one
                    if (numPagesLoadedOnLastGuessGeneration < (dict.Count - guessSet.Count) / 4)
                    {
                        guessGenerator.InvalidateGuess();
                        numPagesLoadedOnLastGuessGeneration = dict.Count - guessSet.Count;
                    }
                    dict.Add(arguments, guessGenerator.GenerateGuess(dict));
                    guessSet.Add(arguments);
                }
                return dict[arguments];
            }
        }


        public ObjectToCache Get(Arguments arguments)
        {
            bool mustLoad;
            lock (dictLock)
            {
                mustLoad = !dict.ContainsKey(arguments) || guessSet.Contains(arguments);
            }

            if (mustLoad)
            {
                //synchronous load

                if (pendingRequests.ContainsKey(arguments))
                {
                    //this request ist already pending, wait for its completion
                    //ObjectToCache res = pendingRequests[arguments].Wait().output;
                    //LoadCompleted(arguments, res);
                    AsyncLoadCompleted(pendingRequests[arguments].Wait(), null);

                }
                else
                {
                    //generate request to load
                    ObjectToCache result = loadObjectAsyncDelegate(arguments).Wait().output;
                    LoadCompleted(arguments, result);
                }
            }

            //asynchronous load (load other elements to cache that are 'close' to the requested one)
            if (predictionAlgorithm != null)
            {
                IList<Arguments> argses = predictionAlgorithm(arguments);
                foreach (Arguments args in argses)
                {
                    loadAsync(args);
                }
            }

            lock (dictLock)
            {
                return dict[arguments];
            }

        }

        private APdfRequest<Arguments, ObjectToCache> loadAsync(Arguments args)
        {
            lock (dictLock)
            {
                if (!pendingRequests.ContainsKey(args) && (!dict.ContainsKey(args) || guessSet.Contains(args)))
                {
                    APdfRequest<Arguments, ObjectToCache> request = loadObjectAsyncDelegate(args);
                    pendingRequests.Add(args, request);
                    request.Completed += AsyncLoadCompleted;
                    return request;
                }
                return pendingRequests.ContainsKey(args) ? pendingRequests[args] : null;
            }
        }

        public void AsyncLoadCompleted(APdfRequest<Arguments, ObjectToCache>.InOutTuple result, PdfViewerCSharpAPI.Utilities.PdfViewerException ex)
        {
            if (ex != null)
                return;

            //only add the loaded values if there is a pending request for it
            if (pendingRequests.ContainsKey(result.arguments))
            {
                //do not load if we have already loaded the same key/value combination
                if(objectEquatorDelegate == null || !dict.ContainsKey(result.arguments) || !objectEquatorDelegate(dict[result.arguments], result.output))
                {
                    LoadCompleted(result.arguments, result.output);
                }
            }
        }
        private void LoadCompleted(Arguments args, ObjectToCache result)
        {
            lock (dictLock)
            {
                if (onFirstLoadParametrizer != null)
                    onFirstLoadParametrizer(result, param);
                guessSet.Remove(args);
                pendingRequests.Remove(args);
                exactlyLoadedSet.Add(args);
                if (dict.ContainsKey(args))
                    dict[args] = result;
                else
                    dict.Add(args, result);
                if (ItemLoadedToCache != null)
                    ItemLoadedToCache(args);
            }
        }

        public void ChangeParametrization(Parametrization newParam)
        {
            if (guessGenerator != null)
                guessGenerator.InvalidateGuess();

            lock (dictLock)
            {
                foreach (ObjectToCache cachedObject in dict.Values)
                {
                    parametrizationChanger(cachedObject, param, newParam);
                }
            }
            this.param = newParam;
        }

        public void InvalidateCache()
        {
            lock (dictLock)
            {
                dict.Clear();
                guessSet.Clear();
                exactlyLoadedSet.Clear();
                pendingRequests.Clear();
                if (guessGenerator != null)
                    guessGenerator.InvalidateGuess();
                numPagesLoadedOnLastGuessGeneration = 0;
            }
        }

        /// <summary>
        /// loads object to cache if necessary asynchronously
        /// </summary>
        /// <param name="arguments">arguments</param>
        /// <returns>null if the element is already loaded, otherwise request</returns>
        public APdfRequest<Arguments, ObjectToCache> GetAsync(Arguments arguments)
        {
            return loadAsync(arguments);
        }

        public void CancelPendingRequest(APdfRequest<Arguments, ObjectToCache> request)
        {
            pendingRequests.Remove(request.InputArguments);
        }

        public bool ExactlyLoaded(Arguments arguments)
        {
            return exactlyLoadedSet.Contains(arguments);
        }

        public event Action<Arguments> ItemLoadedToCache;

        public override string ToString()
        {
            //Collection<IFormattable> list = dict.Keys as ICollection<IFormattable>;
            lock (dict)
            {
                return Utilities.Logger.ListToString<Arguments>(dict.Keys);
            }
        }

        #region set delegates and members
        LoadObjectAsync<Arguments, ObjectToCache> loadObjectAsyncDelegate = null;
        public LoadObjectAsync<Arguments, ObjectToCache> LoadObjectAsyncDelegate
        {
            set
            {
                loadObjectAsyncDelegate = value;
            }
        }
        ObjectEquator<ObjectToCache> objectEquatorDelegate = null;
        public ObjectEquator<ObjectToCache> ObjectEquatorDelegate
        {
            set
            {
                objectEquatorDelegate = value;
            }
        }
        PredictAdditionalObjectsToLoad<Arguments> predictionAlgorithm = null;
        public PredictAdditionalObjectsToLoad<Arguments> PredictionAlgorithm
        {
            set
            {
                predictionAlgorithm = value;
            }
        }
        FirstLoadParametrizer<ObjectToCache, Parametrization> onFirstLoadParametrizer = null;
        public FirstLoadParametrizer<ObjectToCache, Parametrization> FirstLoadParametrizer
        {
            set
            {
                onFirstLoadParametrizer = value;
            }
        }

        ParametrizationChanger<ObjectToCache, Parametrization> parametrizationChanger = null;
        public ParametrizationChanger<ObjectToCache, Parametrization> ParametrizationChanger
        {
            set
            {
                parametrizationChanger = value;
            }
        }

        IGuessGenerator<Arguments, ObjectToCache> guessGenerator = null;
        public IGuessGenerator<Arguments, ObjectToCache> GuessGenerator
        {
            set
            {
                guessGenerator = value;
            }
        }
        #endregion
    }
}

File: Java/PdfViewerAPI_R2/java/com/pdf_tools/pdfviewer/caching/AlwaysRememberCache.java
package com.pdf_tools.pdfviewer.caching;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.pdf_tools.pdfviewer.Model.PdfViewerException;

public class AlwaysRememberCache<Arguments, ObjectToCache, Item, Parametrization>
        implements IGenericCache<Arguments, ObjectToCache, Item, Parametrization>
{

    private Map<Arguments, ObjectToCache> dict;
    private Set<Arguments> guessSet;
    private Set<Arguments> argumentsLoaded;
    private Parametrization param;
    private int numPagesLoadedOnLastGuessGeneration = 0;

    public AlwaysRememberCache()
    {
        dict = new HashMap<Arguments, ObjectToCache>();
        guessSet = new HashSet<Arguments>();
        argumentsLoaded = new HashSet<Arguments>();
        itemLoadedToCacheListenerList = new ArrayList<IItemLoadedToCacheListener<Arguments>>();
    }

    @Override
    public ObjectToCache get(Arguments arguments) throws PdfViewerException
    {

        if (objectLoader == null)
            throw new IllegalStateException("objectLoader of AlwaysRememberCache has not been set");
        List<Arguments> argses;
        if (predictor != null)
        {
            argses = predictor.predictObjectsToLoad(arguments);
        } else
        {
            argses = Collections.singletonList(arguments);
        }

        for (Arguments args : argses)
        {
            if (!dict.containsKey(args) || guessSet.contains(args))
            {
                ObjectToCache result = objectLoader.loadObject(args);
                argumentsLoaded.add(args);
                if (parametrizer != null)
                    parametrizer.firstLoadParametrize(result, param);
                guessSet.remove(args);
                dict.put(args, result);
                for (IItemLoadedToCacheListener<Arguments> listener : itemLoadedToCacheListenerList)
                    listener.onItemLoadedToCache(args);
            }
        }
        return dict.get(arguments);
    }

    @Override
    public ObjectToCache getGuess(Arguments arguments) throws PdfViewerException
    {
        if (objectLoader == null)
            throw new IllegalStateException("objectLoader of AlwaysRememberCache has not been set");
        if (dict.size() <= 0 || guessGenerator == null)
            return get(arguments); // We have to get a real value if the cache
                                   // is empty

        if (!dict.containsKey(arguments))
        {
            // if we have loaded a lot more pages since we last generated a new
            // guess, we should invalidate that old guess and generate a new one
            if (numPagesLoadedOnLastGuessGeneration < (dict.size() - guessSet.size()) / 4)
            {
                guessGenerator.invalidateGuess();
                numPagesLoadedOnLastGuessGeneration = dict.size() - guessSet.size();
            }

            dict.put(arguments, guessGenerator.generateGuess(dict));
            guessSet.add(arguments);
        }
        return dict.get(arguments);
    }

    @Override
    public void invalidateCache()
    {
        if (this.itemUnloader != null)
        {
            for (ObjectToCache obj : dict.values())
                itemUnloader.unloadItem(obj);
        }
        dict.clear();
        guessSet.clear();
        if (guessGenerator != null)
            guessGenerator.invalidateGuess();
        numPagesLoadedOnLastGuessGeneration = 0;
    }

    @Override
    public void changeParametrization(Parametrization newParam)
    {
        if (guessGenerator != null)
            guessGenerator.invalidateGuess();
        for (ObjectToCache element : dict.values())
            parametrizer.changeParametrizationOfLoadedObject(element, param, newParam);
        this.param = newParam;
    }

    @Override
    public void setObjectLoader(com.pdf_tools.pdfviewer.caching.IGenericCache.IObjectLoader<Arguments, ObjectToCache> loader)
    {
        this.objectLoader = loader;
    }

    IObjectLoader<Arguments, ObjectToCache> objectLoader = null;

    @Override
    public void setItemAdder(IGenericCache.IItemAdder<Object, Arguments, Item> adder)
    {
        this.itemAdder = adder;
    }

    @Override
    public void addItemToCache(Arguments arguments, Item item)
    {
        this.itemAdder.add(this.dict, arguments, item);
    }

    IItemAdder<Object, Arguments, Item> itemAdder = null;

    @Override
    public void setItemRemover(IGenericCache.IItemRemover<Object, Arguments, Item> remover)
    {
        this.itemRemover = remover;
    }

    @Override
    public void removeItemFromCache(Arguments args, Item item)
    {
        this.itemRemover.remove(this.dict, args, item);
    }

    IItemRemover<Object, Arguments, Item> itemRemover = null;

    @Override
    public void setArgumentsToLoadPredictor(com.pdf_tools.pdfviewer.caching.IGenericCache.IArgumentsToLoadPredictor<Arguments> predictor)
    {
        this.predictor = predictor;
    }

    IArgumentsToLoadPredictor<Arguments> predictor;

    @Override
    public void setParametrizer(com.pdf_tools.pdfviewer.caching.IGenericCache.IParametrizer<ObjectToCache, Parametrization> parametrizer)
    {
        this.parametrizer = parametrizer;
    }

    IParametrizer<ObjectToCache, Parametrization> parametrizer = null;

    @Override
    public void setGuessGenerator(com.pdf_tools.pdfviewer.caching.IGenericCache.IGuessGenerator<Arguments, ObjectToCache> guessGenerator)
    {
        this.guessGenerator = guessGenerator;
    }

    IGuessGenerator<Arguments, ObjectToCache> guessGenerator = null;

    @Override
    public void addItemLoadedToCacheListener(com.pdf_tools.pdfviewer.caching.IGenericCache.IItemLoadedToCacheListener<Arguments> listener)
    {
        itemLoadedToCacheListenerList.add(listener);
    }

    List<IItemLoadedToCacheListener<Arguments>> itemLoadedToCacheListenerList;

    @Override
    public void setItemUnloader(com.pdf_tools.pdfviewer.caching.IGenericCache.IItemUnloader<ObjectToCache> unloader)
    {
        this.itemUnloader = unloader;
    }

    IItemUnloader<ObjectToCache> itemUnloader = null;

    @Override
    public void unloadObject(Arguments arguments)
    {
        this.dict.remove(arguments);
    }

    @Override
    public Set<Arguments> getArgumentsLoaded()
    {
        return argumentsLoaded;
    }

}
